scipy/stats/stats.py

```diff
@@ -3029,28 +3057,51 @@ def pearsonr(x, y):
     >>> stats.pearsonr([1,2,3,4,5], [5,6,7,8,7])
     (0.83205029433784372, 0.080509573298498519)
     """
-    # x and y should have same length.
+    n = len(x)
+    if n != len(y):
+        raise ValueError('x and y must have the same length.')
+
+    if n < 2:
+        raise ValueError('x and y must have length at least 2.')
+
     x = np.asarray(x)
     y = np.asarray(y)
-    n = len(x)
-    mx = x.mean()
-    my = y.mean()
-    xm, ym = x - mx, y - my
-    r_num = np.add.reduce(xm * ym)
-    r_den = np.sqrt(_sum_of_squares(xm) * _sum_of_squares(ym))
-    r = r_num / r_den
+
+    # If an input is constant, the correlation coefficient is not defined.
+    if (x == x[0]).all() or (y == y[0]).all():
+        warnings.warn(PearsonRConstantInputWarning())
+        return np.nan, np.nan
+
+    if n == 2:
+        return np.float64(np.sign(x[1] - x[0])*np.sign(y[1] - y[0])), 1.0
```

Comment:
This probably should be explained in the documentation as a special case, otherwise user will ask why the p-value is one

Reply:
Agreed.  I'll add an explanation to the docstring.

Reply:
I added an explanation of this to the "Notes" section of the docstring.

Reply:
The `n == 2` case is now explained in the docstring, so marking this as resolved